### PR TITLE
Added location caching; added logging to file

### PR DESCRIPTION
**Pull request by thuryn**
Added caching of locations; added logging to file

### CODE REVIEW

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java [hunk 1]
```diff
@@ -42,11 +44,18 @@
...
-import java.util.Objects;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import static android.os.Build.VERSION.RELEASE;
+import android.os.Parcel;
+import android.os.PowerManager;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
```
Comment (mar-v-in) [inline src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: Please ensure that imports are formatted properly (i.e. ordered by name, one line between different first level package name)

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java [hunk 2]
```diff
@@ -304,4 +408,163 @@ String retrieveString() {
...
+        // Gets the data repository in write mode
+        SQLiteDatabase db = mDbHelper.getWritableDatabase();
+
+        // Create a new map of values, where column names are the keys
+        ContentValues values = new ContentValues();
+        values.put(LocationAddressCache.COLUMN_NAME_ADDRESS, getAddressAsBytes(address));
+        values.put(LocationAddressCache.COLUMN_NAME_LONGITUDE, longitude);
+        values.put(LocationAddressCache.COLUMN_NAME_LATITUDE, latitude);
+        values.put(LocationAddressCache.COLUMN_NAME_LOCALE, locale);
+        values.put(LocationAddressCache.COLUMN_NAME_CREATED, iso8601Format.format(new Date()));
```
Comment (mar-v-in) [inline src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: Please use unix timestamp in seconds or milliseconds for storing timestamps in database.

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java [hunk 3]
```diff
@@ -304,4 +408,163 @@ String retrieveString() {
...
+    }
+
+    private void deleteRecordFromTable(Integer recordId) {
+        SQLiteDatabase db = mDbHelper.getWritableDatabase();
+        String selection = LocationAddressCache._ID + " = ?";
+        String[] selectionArgs = { recordId.toString() };
+        db.delete(LocationAddressCache.TABLE_NAME, selection, selectionArgs);
+    }
+
+    private byte[] getAddressAsBytes(Address address) {
```
Comment (mar-v-in) [inline src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: Would be better in ReverseGeocodingCacheDbHelper next to getAddressFromBytes() (or: put both into ReverseGeocodingCacheContract)

File: src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -4,12 +4,15 @@
           xmlns:tools="http://schemas.android.com/tools">
 
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.WAKE_LOCK" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
```
Comment (mar-v-in) [inline src/main/AndroidManifest.xml:8]: This permission has to be requested on Android 7+

**Review by mar-v-in — COMMENTED** [review]
Do you think it's worth pulling in a third-party dependency and additional permission for a log file feature, that is unlikely to be interesting for the majority of users?

logcat should be sufficient for most debug cases. You could also make logging to file only available in the debug version.

**Comment by thuryn** [comment]
Reply to:
Do you think it's worth pulling in a third-party dependency and additional permission for a log file feature, that is unlikely to be interesting for the majority of users?

logcat should be sufficient for most debug cases. You could also make logging to file only available in the debug version.


I've added the log to file feature because I spent tens of hours debugging in the field. Things can work when you are at home, but it does not when you are outside (run on battery, no wifi, specific netowrk CELLs you don't have in DB, etc.). Unfortunately, logcat is useless for regular user as well as for long time logging. You have to have adb (what regular user does not have), you have to connect phone to PC (what you don't have outside), logcat contains a lot of mess what makes the log unusable.
  So I wanted to give the easiest possibility to user to grab the logs and send it to developer. I can remember the days I wonted to report some bug, but I don't know adb and logcat etc.
  Regarding the perrmissions - I completelly agree with you - the smallest amount of permission is better - you know, I wan't to have INSTALL_PROVIDER permission in UnifiedNLP standalone for the same reason. I cannot see any other possibility. This option is for 1% (or less)  users, but it's important to find all bugs in the service - because newer version of Android usually means that something does not work - because of battery saving etc. And this is something you cannot find at home by computer.

File: src/main/AndroidManifest.xml (hunk 4 above)
Comment (thuryn) [reply src/main/AndroidManifest.xml:8]: fixed

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java (hunk 1 above)
Comment (thuryn) [reply src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: fixed

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java (hunk 2 above)
Comment (thuryn) [reply src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: fixed

File: src/main/java/org/microg/nlp/backend/nominatim/BackendService.java (hunk 3 above)
Comment (thuryn) [reply src/main/java/org/microg/nlp/backend/nominatim/BackendService.java]: fixed

**Comment by ale5000-git** [comment]
@thuryn
Could you please separate location caching and file logging in 2 separate pull requests?
It would simplify reviewing.

**Pull request closed without merging**